File: app/src/main/java/com/todoroo/astrid/service/TaskMover.java
package com.todoroo.astrid.service;

import static com.google.common.collect.FluentIterable.from;
import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.collect.Lists.transform;
import static com.todoroo.andlib.utility.DateUtilities.now;
import static java.util.Collections.emptyList;

import androidx.annotation.Nullable;
import com.todoroo.astrid.api.CaldavFilter;
import com.todoroo.astrid.api.Filter;
import com.todoroo.astrid.api.GtasksFilter;
import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.inject.Inject;
import org.tasks.LocalBroadcastManager;
import org.tasks.data.CaldavDao;
import org.tasks.data.CaldavTask;
import org.tasks.data.GoogleTask;
import org.tasks.data.GoogleTaskDao;
import org.tasks.data.GoogleTaskListDao;
import org.tasks.preferences.Preferences;

public class TaskMover {
  private final TaskDao taskDao;
  private final CaldavDao caldavDao;
  private final GoogleTaskDao googleTaskDao;
  private final GoogleTaskListDao googleTaskListDao;
  private final Preferences preferences;
  private final LocalBroadcastManager localBroadcastManager;

  @Inject
  public TaskMover(
      TaskDao taskDao,
      CaldavDao caldavDao,
      GoogleTaskDao googleTaskDao,
      GoogleTaskListDao googleTaskListDao,
      Preferences preferences,
      LocalBroadcastManager localBroadcastManager) {
    this.taskDao = taskDao;
    this.caldavDao = caldavDao;
    this.googleTaskDao = googleTaskDao;
    this.googleTaskListDao = googleTaskListDao;
    this.preferences = preferences;
    this.localBroadcastManager = localBroadcastManager;
  }

  public Filter getSingleFilter(List<Long> tasks) {
    List<String> caldavCalendars = caldavDao.getCalendars(tasks);
    List<String> googleTaskLists = googleTaskDao.getLists(tasks);
    if (caldavCalendars.isEmpty()) {
      if (googleTaskLists.size() == 1) {
        return new GtasksFilter(googleTaskListDao.getByRemoteId(googleTaskLists.get(0)));
      }
    } else if (googleTaskLists.isEmpty()) {
      if (caldavCalendars.size() == 1) {
        return new CaldavFilter(caldavDao.getCalendar(caldavCalendars.get(0)));
      }
    }
    return null;
  }

  public void move(List<Long> tasks, Filter selectedList) {
    tasks = new ArrayList<>(tasks);
    tasks.removeAll(googleTaskDao.findChildrenInList(tasks));
    tasks.removeAll(taskDao.findChildrenInList(tasks));
    taskDao.setParent(0, null, tasks);
    for (Task task : taskDao.fetch(tasks)) {
      performMove(task, selectedList);
    }
    if (selectedList instanceof CaldavFilter) {
      caldavDao.updateParents((((CaldavFilter) selectedList).getUuid()));
    }
    taskDao.touch(tasks);
    localBroadcastManager.broadcastRefresh();
  }

  private void performMove(Task task, @Nullable Filter selectedList) {
    long id = task.getId();

    GoogleTask googleTask = googleTaskDao.getByTaskId(id);
    if (googleTask != null) {
      moveGoogleTask(task, googleTask, selectedList);
      return;
    }

    CaldavTask caldavTask = caldavDao.getTask(id);
    if (caldavTask != null) {
      moveCaldavTask(task, caldavTask, selectedList);
      return;
    }

    moveLocalTask(task, selectedList);
  }

  private void moveGoogleTask(Task task, GoogleTask googleTask, Filter selected) {
    if (selected instanceof GtasksFilter
        && googleTask.getListId().equals(((GtasksFilter) selected).getRemoteId())) {
      return;
    }

    long id = googleTask.getTask();
    List<GoogleTask> children = googleTaskDao.getChildren(id);
    List<Long> childIds = from(children).transform(GoogleTask::getTask).toList();
    googleTaskDao.markDeleted(now(), id);

    if (selected instanceof GtasksFilter) {
      String listId = ((GtasksFilter) selected).getRemoteId();
      googleTaskDao.insertAndShift(new GoogleTask(id, listId), preferences.addTasksToTop());
      if (!children.isEmpty()) {
        googleTaskDao.insert(
            transform(
                children,
                child -> {
                  GoogleTask newChild = new GoogleTask(child.getTask(), listId);
                  newChild.setOrder(child.getOrder());
                  newChild.setParent(id);
                  return newChild;
                }));
      }
    } else if (selected instanceof CaldavFilter) {
      String listId = ((CaldavFilter) selected).getUuid();
      CaldavTask newParent = new CaldavTask(id, listId);
      caldavDao.insert(task, newParent, preferences.addTasksToTop());
      caldavDao.insert(
          transform(
              childIds,
              child -> {
                CaldavTask newChild = new CaldavTask(child, listId);
                newChild.setRemoteParent(newParent.getRemoteId());
                return newChild;
              }));
    } else {
      taskDao.setParent(task.getId(), task.getUuid(), childIds);
    }
  }

  private void moveCaldavTask(Task task, CaldavTask caldavTask, Filter selected) {
    if (selected instanceof CaldavFilter
        && caldavTask.getCalendar().equals(((CaldavFilter) selected).getUuid())) {
      return;
    }

    long id = task.getId();
    List<Long> childIds = taskDao.getChildren(id);
    List<Long> toDelete = newArrayList(id);
    List<CaldavTask> children = emptyList();
    if (!childIds.isEmpty()) {
      children = caldavDao.getTasks(childIds);
      toDelete.addAll(childIds);
    }
    caldavDao.markDeleted(now(), toDelete);

    if (selected instanceof CaldavFilter) {
      long id1 = caldavTask.getTask();
      String listId = ((CaldavFilter) selected).getUuid();
      CaldavTask newParent =
          new CaldavTask(id1, listId, caldavTask.getRemoteId(), caldavTask.getObject());
      newParent.setVtodo(caldavTask.getVtodo());
      caldavDao.insert(task, newParent, preferences.addTasksToTop());
      caldavDao.insert(
          transform(
              children,
              child -> {
                CaldavTask newChild =
                    new CaldavTask(child.getTask(), listId, child.getRemoteId(), child.getObject());
                newChild.setVtodo(child.getVtodo());
                newChild.setRemoteParent(child.getRemoteParent());
                return newChild;
              }));
    } else if (selected instanceof GtasksFilter) {
      moveToGoogleTasks(id, childIds, (GtasksFilter) selected);
    } else {
      taskDao.updateParentUids(from(children).transform(CaldavTask::getTask).toList());
    }
  }

  private void moveLocalTask(Task task, @Nullable Filter selected) {
    if (selected instanceof GtasksFilter) {
      moveToGoogleTasks(task.getId(), taskDao.getChildren(task.getId()), (GtasksFilter) selected);
    } else if (selected instanceof CaldavFilter) {
      long id = task.getId();
      String listId = ((CaldavFilter) selected).getUuid();
      Map<Long, CaldavTask> tasks = new HashMap<>();
      CaldavTask root = new CaldavTask(id, listId);
      for (Task child : taskDao.fetchChildren(task.getId())) {
        CaldavTask newTask = new CaldavTask(child.getId(), listId);
        long parent = child.getParent();
        newTask.setRemoteParent((parent == id ? root : tasks.get(parent)).getRemoteId());
        tasks.put(child.getId(), newTask);
      }
      caldavDao.insert(task, root, preferences.addTasksToTop());
      caldavDao.insert(tasks.values());
    }
  }

  private void moveToGoogleTasks(long id, List<Long> children, GtasksFilter filter) {
    taskDao.setParent(0, null, children);
    String listId = filter.getRemoteId();
    googleTaskDao.insertAndShift(new GoogleTask(id, listId), preferences.addTasksToTop());
    List<GoogleTask> newChildren = new ArrayList<>();
    for (int i = 0; i < children.size(); i++) {
      GoogleTask newChild = new GoogleTask(children.get(i), listId);
      newChild.setOrder(i);
      newChild.setParent(id);
      newChildren.add(newChild);
    }
    googleTaskDao.insert(newChildren);
  }
}

File: app/src/main/java/com/todoroo/astrid/service/TaskMover.kt
package com.todoroo.astrid.service

import org.tasks.LocalBroadcastManager
import org.tasks.caldav.VtodoCache
import org.tasks.data.dao.CaldavDao
import org.tasks.data.dao.GoogleTaskDao
import org.tasks.data.dao.TaskDao
import org.tasks.data.db.DbUtils.dbchunk
import org.tasks.data.entity.CaldavTask
import org.tasks.data.entity.Task
import org.tasks.data.getLocalList
import org.tasks.filters.CaldavFilter
import org.tasks.filters.Filter
import org.tasks.preferences.Preferences
import org.tasks.sync.SyncAdapters
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import javax.inject.Inject

class TaskMover @Inject constructor(
    private val taskDao: TaskDao,
    private val caldavDao: CaldavDao,
    private val googleTaskDao: GoogleTaskDao,
    private val preferences: Preferences,
    private val localBroadcastManager: LocalBroadcastManager,
    private val syncAdapters: SyncAdapters,
    private val vtodoCache: VtodoCache,
) {

    suspend fun getSingleFilter(tasks: List<Long>): Filter? {
        val caldavCalendars = caldavDao.getCalendars(tasks)
        return if (caldavCalendars.size == 1) {
            val list = caldavCalendars.first()
            val account = list.account?.let { caldavDao.getAccountByUuid(it) }
            account?.let { CaldavFilter(calendar = list, account = it) }
        } else {
            null
        }
    }

    suspend fun move(task: Long, list: Long) {
        val calendar = caldavDao.getCalendarById(list) ?: return
        val account = calendar.account?.let { caldavDao.getAccountByUuid(it) } ?: return
        move(
            ids = listOf(task),
            selectedList = CaldavFilter(calendar = calendar, account = account),
        )
    }

    suspend fun move(ids: List<Long>, selectedList: CaldavFilter) {
        val tasks = ids
            .dbchunk()
            .flatMap { taskDao.getChildren(it) }
            .let { taskDao.fetch(ids.minus(it.toSet())) }
            .filterNot { it.readOnly }
        val taskIds = tasks.map { it.id }
        taskDao.setParent(0, ids.intersect(taskIds.toSet()).toList())
        tasks.forEach { performMove(it, selectedList) }
        if (!selectedList.isGoogleTasks) {
            caldavDao.updateParents(selectedList.uuid)
        }
        taskIds.dbchunk().forEach {
            taskDao.touch(it)
        }
        localBroadcastManager.broadcastRefresh()
        syncAdapters.sync()
    }

    suspend fun migrateLocalTasks() {
        val list = caldavDao.getLocalList()
        val account = list.account?.let { caldavDao.getAccountByUuid(it) } ?: return
        move(taskDao.getLocalTasks(), CaldavFilter(calendar = list, account = account))
    }

    private suspend fun performMove(task: Task, selectedList: CaldavFilter) {
        googleTaskDao.getByTaskId(task.id)?.let {
            moveGoogleTask(task, it, selectedList)
            return
        }
        caldavDao.getTask(task.id)?.let {
            moveCaldavTask(task, it, selectedList)
            return
        }
        moveLocalTask(task, selectedList)
    }

    private suspend fun moveGoogleTask(task: Task, googleTask: CaldavTask, selected: CaldavFilter) {
        if (googleTask.calendar == selected.uuid) {
            return
        }
        val id = task.id
        val children = taskDao.getChildren(id)
        caldavDao.markDeleted(children + id, currentTimeMillis())
        when {
            selected.isGoogleTasks -> {
                val listId = selected.uuid
                googleTaskDao.insertAndShift(
                    task = task,
                    caldavTask = CaldavTask(
                        task = id,
                        calendar = listId,
                        remoteId = null,
                    ),
                    top = preferences.addTasksToTop()
                )
                children.takeIf { it.isNotEmpty() }
                        ?.map {
                            CaldavTask(
                                task = it,
                                calendar = listId,
                                remoteId = null,
                            )
                        }
                        ?.let { googleTaskDao.insert(it) }
            }
            else -> {
                val listId = selected.uuid
                val newParent = CaldavTask(
                    task = id,
                    calendar = listId,
                )
                caldavDao.insert(task, newParent, preferences.addTasksToTop())
                children.map {
                    val newChild = CaldavTask(
                        task = it,
                        calendar = listId
                    )
                    newChild.remoteParent = newParent.remoteId
                    newChild
                }.let { caldavDao.insert(it) }
            }
        }
    }

    private suspend fun moveCaldavTask(task: Task, caldavTask: CaldavTask, selected: CaldavFilter) {
        if (caldavTask.calendar == selected.uuid) {
            // TODO: make sure its the same account
            return
        }
        val id = task.id
        val childIds = taskDao.getChildren(id)
        val toDelete = arrayListOf(id)
        var children: List<CaldavTask> = emptyList()
        if (childIds.isNotEmpty()) {
            children = caldavDao.getTasks(childIds)
            toDelete.addAll(childIds)
        }
        caldavDao.markDeleted(toDelete, currentTimeMillis())
        when {
            selected.isGoogleTasks -> moveToGoogleTasks(id, childIds, selected)
            else -> {
                val from = caldavDao.getCalendar(caldavTask.calendar!!)
                val id1 = caldavTask.task
                val listId = selected.uuid
                val newParent = CaldavTask(
                    task = id1,
                    calendar = listId,
                    remoteId = caldavTask.remoteId,
                    obj = caldavTask.obj,
                )
                vtodoCache.move(from!!, selected.calendar, caldavTask)
                caldavDao.insert(task, newParent, preferences.addTasksToTop())
                children.takeIf { it.isNotEmpty() }
                        ?.map {
                            val newChild = CaldavTask(
                                task = it.task,
                                calendar = listId,
                                remoteId = it.remoteId,
                                obj = it.obj,
                            )
                            vtodoCache.move(from, selected.calendar, it)
                            newChild.remoteParent = it.remoteParent
                            newChild
                        }
                        ?.let { caldavDao.insert(it) }
            }
        }
    }

    private suspend fun moveLocalTask(task: Task, selected: CaldavFilter) {
        when {
            selected.isGoogleTasks -> moveToGoogleTasks(task.id, taskDao.getChildren(task.id), selected)
            else -> {
                val id = task.id
                val listId = selected.uuid
                val tasks: MutableMap<Long, CaldavTask> = HashMap()
                val root = CaldavTask(
                    task = id,
                    calendar = listId,
                )
                val children = taskDao.getChildren(id).mapNotNull { taskDao.fetch(it) }
                for (child in children) {
                    val newTask = CaldavTask(
                        task = child.id,
                        calendar = listId,
                    )
                    val parent = child.parent
                    newTask.remoteParent = (if (parent == id) root else tasks[parent])!!.remoteId
                    tasks[child.id] = newTask
                }
                caldavDao.insert(task, root, preferences.addTasksToTop())
                caldavDao.insert(tasks.values)
            }
        }
    }

    private suspend fun moveToGoogleTasks(id: Long, children: List<Long>, filter: CaldavFilter) {
        if (!filter.isGoogleTasks) {
            // TODO: make sure its the same account
            return
        }
        val task = taskDao.fetch(id) ?: return
        taskDao.setParent(id, children)
        val listId = filter.uuid
        googleTaskDao.insertAndShift(
            task,
            CaldavTask(
                task = id,
                calendar = listId,
                remoteId = null
            ),
            preferences.addTasksToTop()
        )
        children.takeIf { it.isNotEmpty() }
                ?.map {
                    CaldavTask(
                        task = it,
                        calendar = listId,
                        remoteId = null
                    )
                }
                ?.let { googleTaskDao.insert(it) }
    }
}